### PR TITLE
Fix Windows compilation of test sources

### PR DESCRIPTION
**Pull request by wrowe**
Description:

This patch is extracted from https://github.com/envoyproxy/envoy/pull/10542 and must be preceded by that merge in order for all enabled tests to compile successfully on Windows.

This patch should still be correct in the interim, and should effectively be a no-op on Linux.

- Remove erroneous test from PR #9964
  - Test macro missing the second argument (for test name),
    causing compilation on Windows to fail. It appears to
    have no purpose, copy-paste error?
- more involved compilation failures in test sources are
  excluded with skip_on_windows tag
  - involving tests to do with features that don't exist for
    windows (signals, etc.)
  - sh_test rules do not currently work on Windows, waiting on a Bazel
    fix
- Individually exclude tests in non-compiling extensions
  with skip_on_windows
- subset of failing tests tagged with fails_on_windows
  - tests that time out
  - tests that are known to fail
  - non-exhaustive list, additional failing tests will be
    fixed/tagged in subsequent efforts
- various fixes for Windows compilation
  - fix os_fd_t type in mocks
  - ensure consistent integer types
  - proliferate use of OsSysCalls in tests
  - exclude signals tests
  - add _set_invalid_parameter_handler to test main to ensure runtime
    does not abort when functions are called with invalid parameters
  - print warning about tcpdump not currently being enabled on Windows
  - matchers_test.cc: do not return address of literal
  - test/extensions/filters/network/direct_response/BUILD: skip test with
    constructor+lambda that does not compile on Windows
  - codec_impl_test.cc: MSVC pre-processing does not like the multline
    string passed to EXPECT_EQ macro, pull into intermediate variable

Co-authored-by: Sunjay Bhatia <sbhatia@pivotal.io>
Co-authored-by: William A Rowe Jr <wrowe@pivotal.io>
Signed-off-by: Sunjay Bhatia <sbhatia@pivotal.io>
Signed-off-by: William A Rowe Jr <wrowe@pivotal.io>

Risk Level: Low
Testing: Local on msvc-cl, linux-gcc
Docs Changes: n/a
Release Notes: n/a
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by wrowe** [comment]
Ready for review and merge. As we noted in PR 10542;

> Before removing test/* changes from this PR, all tests passing, except for coverage (known to be unstable.)

Any coverage and macos regressions indicate new problems on master or issues with the CI build environments themselves, not this specific patch.

File: test/common/buffer/owned_impl_test.cc [hunk 1]
```diff
@@ -649,19 +649,24 @@ TEST_F(OwnedImplTest, ReserveZeroCommit) {
     slices[i].len_ = 0;
   }
   buf.commit(slices, allocated_slices);
-  int pipe_fds[2] = {0, 0};
-  ASSERT_EQ(::pipe(pipe_fds), 0);
+  os_fd_t pipe_fds[2] = {0, 0};
+  auto& os_sys_calls = Api::OsSysCallsSingleton::get();
+#ifdef WIN32
+  ASSERT_EQ(os_sys_calls.socketpair(AF_INET, SOCK_STREAM, 0, pipe_fds).rc_, 0);
```
Comment (yanavlasov) [inline test/common/buffer/owned_impl_test.cc:655]: Should we use socketpair for linux tests as well and avoid ifdefing?

Comment (wrowe) [reply test/common/buffer/owned_impl_test.cc:655]: That is a question of whether we are trying to test tcp or unix domain sockets in the linux implementation. Even though we have resolved that we can have AF_UNIX sockets on windows, a nameless pipe still isn't available. We can stub that in later, but this is the shortest-path to a working test compilation.

So if there are a couple +1's to switching this to os_sys_calls.socketpair, we are fine with that, or we can leave it as-is. A very final step of this integration will be revisiting all of the ifdef/defined() WIN32 to evaluate whether or not they should remain (some of them predate our involvement.)

File: test/extensions/filters/network/mysql_proxy/mysql_integration_test.cc [hunk 2]
```diff
@@ -1,5 +1,8 @@
+#ifndef WIN32
 #include <pthread.h>
```
Comment (mattklein123) [inline test/extensions/filters/network/mysql_proxy/mysql_integration_test.cc:2]: Is this actually needed on linux? I kind of doubt it. Can it just be removed?

File: test/main.cc [hunk 3]
```diff
@@ -13,10 +13,32 @@
...
 using bazel::tools::cpp::runfiles::Runfiles;
 // The main entry point (and the rest of this file) should have no logic in it,
 // this allows overriding by site specific versions of main.cc.
 int main(int argc, char** argv) {
+#if defined(WIN32)
+  _set_abort_behavior(0, _WRITE_ABORT_MSG | _CALL_REPORTFAULT);
+
+  _set_invalid_parameter_handler(noop_invalid_parameter_handler);
+
+  WORD wVersionRequested;
```
Comment (mattklein123) [inline test/main.cc:33]: nit: snake case var names here and below.

File: test/common/buffer/owned_impl_test.cc (hunk 1 above)
Comment (mattklein123) [reply test/common/buffer/owned_impl_test.cc:655]: I think as all of these tests are buffer tests and not network tests, having them all use socket pair seems fine to me so let's just do that?

File: test/extensions/filters/network/mysql_proxy/mysql_integration_test.cc (hunk 2 above)
Comment (wrowe) [reply test/extensions/filters/network/mysql_proxy/mysql_integration_test.cc:2]: Confirmed on gcc.

File: test/common/buffer/owned_impl_test.cc (hunk 1 above)
Comment (wrowe) [reply test/common/buffer/owned_impl_test.cc:655]: Tracking this suggestion in issue 10871, which you can assign back to myself and @sunjayBhatia.

We would appreciate accepting this PR as-is, and we will revisit this after resolving the rest of the Windows CI //test/... RBE changes.

Comment (wrowe) [reply test/common/buffer/owned_impl_test.cc:655]: (I had forgotten that socketpair on Linux fails with AF_INET, so the Windows logic can't be used until we support AF_UNIX on Windows.)

**Comment by wrowe** [comment]
/retest 

... likely won't retest for github 500's overload on the envoy-presubmit pipeline

File: test/common/buffer/owned_impl_test.cc (hunk 1 above)
Comment (mattklein123) [reply test/common/buffer/owned_impl_test.cc:655]: Sure SGTM

**Pull request merged**